sdks/python/apache_beam/io/vcfio.py

```diff
@@ -0,0 +1,380 @@
+        specified in the VCF record and the type of the value is specified by
+        the VCF header FORMAT.
+    """
+    self.name = name
+    self.genotype = genotype or []
+    self.phaseset = phaseset
+    self.info = info or {}
+
+  def __eq__(self, other):
+    return (self.name == other.name and
+            self.genotype == other.genotype and
+            self.phaseset == other.phaseset and
+            self.info == other.info)
+
+  def __repr__(self):
+    return ', '.join(
+        [str(s) for s in [self.name, self.genotype, self.phaseset, self.info]])
+
+
+class _VcfSource(filebasedsource.FileBasedSource):
+  r"""A source for reading VCF files.
+
+  Parses VCF files (version 4) using PyVCF library. If file_pattern specifies
+  multiple files, then the header from each file is used separately to parse
+  the content. However, the output will be a uniform PCollection of
+  :class:`Variant` objects.
+  """
+
+  DEFAULT_VCF_READ_BUFFER_SIZE = 65536  # 64kB
+
+  def __init__(self,
+               file_pattern,
+               compression_type=CompressionTypes.AUTO,
+               buffer_size=DEFAULT_VCF_READ_BUFFER_SIZE,
+               validate=True):
+    super(_VcfSource, self).__init__(file_pattern,
+                                     compression_type=compression_type,
+                                     validate=validate)
+
+    self._header_lines_per_file = {}
+    self._file_pattern = file_pattern
+    self._compression_type = compression_type
+    self._buffer_size = buffer_size
+    self._validate = validate
+
+  def read_records(self, file_name, range_tracker):
+    def line_generator():
+      header_lines = []
+
+      def store_header_lines(header):
+        for line in header:
+          header_lines.append(line)
+
+      text_source = TextSource(
+          self._file_pattern,
+          0,  # min_bundle_size
+          self._compression_type,
+          strip_trailing_newlines=True,
+          coder=coders.StrUtf8Coder(),
+          buffer_size=self._buffer_size,
+          validate=self._validate,
+          skip_header_lines=0,
+          header_matcher_predicate=lambda x: x.startswith('#'),
+          header_processor=store_header_lines)
+
+      records = text_source.read_records(file_name, range_tracker)
+      header_processed = False
+      for line in records:
+        if not header_processed and header_lines:
+          for header in header_lines:
+            yield header
+          header_processed = True
+        # PyVCF has explicit str() calls when parsing INFO fields, which fails
+        # with UTF-8 decoded strings. Encode the line back to UTF-8.
+        yield line.encode('utf-8')
+
+    try:
+      vcf_reader = vcf.Reader(fsock=line_generator())
+      while True:
+        record = next(vcf_reader)
+        yield self._convert_to_variant_record(
+            record, vcf_reader.infos, vcf_reader.formats)
+    except SyntaxError as e:
+      raise ValueError('Invalid VCF header: %s' % str(e))
```

Comment:
<!--thread_id:cc_145273017_t; commit:70ade9d4fe6a8e31ff0be4d39b731510c93b4bf9; resolved:1-->
<!--section:context-quote-->
> **chamikaramj** wrote:
> Are we sure that a 'SyntaxError' only gets raised for invalid headers ? Please add a comment on this if this is the case, otherwise please update the error message to be more generic or let this method simply raise SyntaxError.

<!--section:body-->
Done.